File: GameShelf/Models/GameCollection.swift
//
//  GameCollection.swift
//  GameShelf
//

import Foundation
import SwiftData

/// Una carpeta creada por el usuario para agrupar juegos.
///
/// Es lo que ninguna tienda permite: juntar en un mismo grupo juegos de Steam,
/// PSN y Epic. Un juego puede estar en varias colecciones a la vez.
///
/// El tipo se llama `GameCollection` y no `Collection` a proposito: ese nombre
/// sombrea el protocolo `Swift.Collection` y rompe cualquier codigo generico
/// que lo use (`'Element' is not a member type of type 'C'`).
@Model
final class GameCollection {
  var id: UUID = UUID()

  var name: String = ""

  /// Nombre de un simbolo del sistema (SF Symbols), por ejemplo `folder`.
  var symbolName: String = GameCollection.defaultSymbol

  /// Color de la coleccion. Se guarda como texto por lo mismo que los demas
  /// enums del proyecto.
  var color: CollectionColor = CollectionColor.default

  /// Posicion en la lista. El usuario las reordena arrastrando.
  var sortOrder: Int = 0

  var createdAt: Date = Date()

  /// Juegos que contiene.
  ///
  /// La regla de borrado es `nullify`, la que trae SwiftData por defecto en las
  /// relaciones a varios: borrar una coleccion **no borra los juegos**, solo
  /// deshace la agrupacion.
  @Relationship(inverse: \Game.collections)
  var games: [Game] = []

  static let defaultSymbol = "folder"

  init(
    name: String,
    symbolName: String = GameCollection.defaultSymbol,
    color: CollectionColor = .default,
    sortOrder: Int = 0
  ) {
    self.id = UUID()
    self.name = name
    self.symbolName = symbolName
    self.color = color
    self.sortOrder = sortOrder
    self.createdAt = Date()
    self.games = []
  }

  /// Cuantos juegos tiene.
  var gameCount: Int { games.count }

  /// Si no tiene ningun juego.
  var isEmpty: Bool { games.isEmpty }

  /// Horas jugadas sumando todos sus juegos.
  var totalPlaytimeHours: Double {
    games.reduce(0) { $0 + $1.playtimeHours }
  }

  /// Agrega un juego, sin duplicarlo si ya estaba.
  func add(_ game: Game) {
    guard !contains(game) else { return }
    games.append(game)
  }

  /// Quita un juego de la coleccion. El juego no se borra.
  func remove(_ game: Game) {
    games.removeAll { $0.id == game.id }
  }

  /// Si el juego ya esta en la coleccion.
  func contains(_ game: Game) -> Bool {
    games.contains { $0.id == game.id }
  }
}

File: android/app/src/main/java/com/gameshelf/domain/GameCollection.kt
package com.gameshelf.domain

import java.time.Instant
import java.util.UUID

/**
 * Una carpeta creada por el usuario para agrupar juegos.
 *
 * Es lo que ninguna tienda permite: juntar en un mismo grupo juegos de Steam,
 * PSN y Epic. Un juego puede estar en varias colecciones a la vez.
 *
 * El tipo se llama `GameCollection` y no `Collection` por la misma razon que
 * en la version de iOS: ahi sombreaba `Swift.Collection`, y aca sombrearia
 * `kotlin.collections.Collection`.
 *
 * Borrar una coleccion **no borra los juegos**, solo deshace la agrupacion:
 * en Room eso es un `ON DELETE CASCADE` sobre la tabla puente, no sobre los
 * juegos.
 */
data class GameCollection(
  val id: UUID = UUID.randomUUID(),

  val name: String = "",

  /** Nombre del icono, como valor de dominio. La UI lo traduce a un vector. */
  val symbol: CollectionSymbol = DEFAULT_SYMBOL,

  /** Color de la coleccion. */
  val color: CollectionColor = CollectionColor.DEFAULT,

  /** Posicion en la lista. El usuario las reordena arrastrando. */
  val sortOrder: Int = 0,

  val createdAt: Instant = Instant.now(),

  /**
   * Juegos que contiene.
   *
   * Puede llegar vacia cuando la coleccion viaja dentro de un [Game]: ver la
   * nota sobre el ciclo en la documentacion de [Game].
   */
  val games: List<Game> = emptyList(),
) {
  /** Cuantos juegos tiene. */
  val gameCount: Int get() = games.size

  /** Si no tiene ningun juego. */
  val isEmpty: Boolean get() = games.isEmpty()

  /** Horas jugadas sumando todos sus juegos. */
  val totalPlaytimeHours: Double get() = games.sumOf { it.playtimeHours }

  /** Agrega un juego, sin duplicarlo si ya estaba. */
  fun add(game: Game): GameCollection =
    if (contains(game)) this else copy(games = games + game)

  /** Quita un juego de la coleccion. El juego no se borra. */
  fun remove(game: Game): GameCollection =
    copy(games = games.filterNot { it.id == game.id })

  /** Si el juego ya esta en la coleccion. */
  fun contains(game: Game): Boolean = games.any { it.id == game.id }

  companion object {
    val DEFAULT_SYMBOL = CollectionSymbol.FOLDER
  }
}

/**
 * Iconos que puede llevar una coleccion.
 *
 * En iOS esto era un nombre libre de SF Symbol. Aca es un enum cerrado: los
 * iconos de Material se referencian por propiedad, no por cadena, asi que un
 * nombre guardado que ya no exista no puede resolverse en tiempo de ejecucion.
 * Cerrar el conjunto convierte ese fallo en imposible.
 */
enum class CollectionSymbol(val id: String) {
  FOLDER("folder"),
  STAR("star"),
  HEART("heart"),
  BOOKMARK("bookmark"),
  FLAG("flag"),
  TROPHY("trophy"),
  GAMEPAD("gamepad"),
  CLOCK("clock"),
  SPARKLES("sparkles"),
  LIST("list");

  companion object {
    /** Busca por id guardado; si no se reconoce, cae al icono por defecto. */
    fun fromId(id: String): CollectionSymbol =
      entries.firstOrNull { it.id == id } ?: FOLDER
  }
}
